README.md

```diff
@@ -1,16 +1,1837 @@
+      },
+      "auth": {
+        "/": "bafyreifw2hxvz66fxx7vsn6q3ayrc62smhvt7jta3ck2u6kf4xgqo524hm"
+      },
+      "cause": {
+        "/": "bafyreiekrhjlfend2mwcmtakwshuctotwqcm2wriya6thpjcgvjzfmkt24"
+      },
+      "prf": [
+        {
+          "/": "bafyreieynwqrabzdhgl652ftsk4mlphcj3bxchkj2aw5eb6dc2wxieilau"
+        }
+      ]
+    },
+    "bafyreifw2hxvz66fxx7vsn6q3ayrc62smhvt7jta3ck2u6kf4xgqo524hm": {
+      "scope": [
+        {
+          "/": "bafyreifcerdvicarktlnif5uj25ultgpedwg63nxhmp7anoepaamqj4eji"
+        }
+      ],
+      "s": {
+        "/": {
+          "bytes": "7aEDQEkocMddH2BF4wuMRSlJaJFhu7pQaAInuzLrfU3TQd9XV/Rd0I6N6OvxPaCebNfpuCXrAgtlAxVh+sn8gWSSegk"
+        }
+      }
+    }
+  },
+  "roots": [
+    {
+      "/": "bafyreicc3rilzaxravqju766yr4v7p6nm6sjsd4lbkfa4uqohbbxaejh7e"
+    }
+  ]
+}
+```
+
+# 6 Result
+
+A `Result` records the output of the [Task], as well as its success or failure state.
+
+## 6.1 Schema
+
+```ipldsch
+type Result union {
+  | any "ok"
+  | any "error"
+} representation keyed
+}
+```
+
+## 6.2 Variants
+
+## 6.2.1 Success
+
+The success branch MUST contain the value returned from a successful [Task] wrapped in the `"ok"` tag. The exact shape of the returned data is left undefined to allow for flexibility in various Task types.
+
+```json
+{ "ok": 42 }
+```
+
+## 6.2.2 Failure
+
+The failure branch MAY contain detail about why execution failed wrapped in the "error" tag. It is left undefined in this specification to allow for [Task] types to standardize the data that makes sense in their contexts.
+
+If no information is available, this field SHOULD be set to `{}`.
+
+```json
+{
+  "error": {
+    "dev/reason": "unauthorized",
+    "http/status": 401
+  }
+}
+```
+
+# 7 Effect
+
+Workflows often consist of many, sometimes concurrent, steps which form an execution threads. Steps of such workflows are discrete [Task] [Invocation]s, each producing some [Result] and a description of subsequent step(s) caused by it. Subsequent steps are denoted with an `Effect`, a set of [Invocation]s that [Executor] must perform to advance workflow execution.
+
+All of the [Invocation]s of the `Effect` MUST be concurrent with one another, unless explicitly arranged differently using [Pipelines].
+
+Effect MAY instruct the [Executor] to run set of [Task] [Invocation]s in concurrent execution threads by providing them as list under `fork` field.
+
+Effect MAY instruct the [Executor] that the [Task] [Invocation] is a continuation of the current execution thread by providing it under `join` field.
+
+Often [Invocation] in the `join` field will synthesize [Result]s of the concurrent execution threads (spawned by `fork`), and incorporate them into the current execution thread.
+
+It is also possible for the thread to complete execution (omit `join` field) and leave concurrent threads behind. The `join` field, simply marks the execution thread and provides a way to follow it through trail of [Receipt]s.
```

Comment:
Oh yeah, we're thinking about effects very differently 😅 Not that you _shouldn't_ be able to express this, but how's this for more general wording:

```suggestion
The result of an [Invocation] MAY include a request for further actions to be performed via the `fx` (Effects) field. This enables several things: a clean separation of pure return values from effects, and allows the runtime to decide how (or if!) further work should be performed.

Effects describe requests for future work to be performed. All [Invocation]s generated in an `fx` block MUST be treated as concurrent, unless explicit data dependencies between them exist via promise [Pipelines]. The `fx` block contains two fields: `fork` and `join`.

[Task]s listed in the `fork` field are first-class and only ordered by promises; they otherwise SHOULD be considered independent and equal.

The `join` field describes an OPTIONAL "special" [Invocation] that is treated as a "main" thread. Effects MAY instruct the [Executor] that the [Task] [Invocation] is a continuation of the previous Invocation by providing it under `join` field. This roughly emulates a "main" virtual thread.
```

Reply:
Thanks for the suggested improvement. My only concern with suggested version is it's emphasis of "main" thread. I was describing it as marker of the execution thread instead, because forked invocations could also produce effects with joins of their own emulating multiple concurrent threads of execution and not all of them could be main.